Networking: introduce packet-object based API

apple502j opened the pull request:
See the javadoc for designs. Huge thanks to @Technici4n and @Sollace.

## FAQ
- **Is this a breaking change?** Nope.
- **Is the old API deprecated?** Nope.

apple502j committed:
Initial design of new networking API

apple502j committed:
Add javadoc

apple502j committed:
Deprecate old API

apple502j committed:
Checkstyle

apple502j committed:
Fix testmods

apple502j committed:
Add MinecraftClient param since CPE#client is protected

apple502j committed:
Revert "Deprecate old API"

This reverts commit d3c16efc6526ac8ef5aff0b5e20eb2db60bdf562.

apple502j committed:
Add a note on newer API

apple502j committed:
Add introduction javadoc

Rasmion commented:
Perhaps this should be extended not only to play networking, but also to the login?

apple502j commented:
@RasmusHD Nope, in fact these APIs need a huge rework. I don't think login stuff would live very long.

Juuxel commented on fabric-networking-api-v1/src/main/java/net/fabricmc/fabric/api/networking/v1/PacketType.java:
```diff
@@ -0,0 +1,73 @@
+	}
+
+	@ApiStatus.Internal
+	public T read(PacketByteBuf buf) {
```
Why is this internal? I don't see any harm in exposing this.

Juuxel commented on fabric-networking-api-v1/src/client/java/net/fabricmc/fabric/api/client/networking/v1/ClientPlayNetworking.java:
```diff
@@ -255,4 +403,44 @@ public interface PlayChannelHandler {
+	 * An internal packet handler that works as a proxy between old and new API.
+	 * @param <T> the type of the packet
+	 */
+	interface PlayChannelHandlerProxy<T extends FabricPacket> extends PlayChannelHandler {
```
```suggestion
	private interface PlayChannelHandlerProxy<T extends FabricPacket> extends PlayChannelHandler {
```
I think this can be private?

Juuxel commented on fabric-networking-api-v1/src/client/java/net/fabricmc/fabric/api/client/networking/v1/ClientPlayNetworking.java:
```diff
@@ -255,4 +403,44 @@ public interface PlayChannelHandler {
+		 * <p>An example usage of this is to display an overlay message:
+		 * <pre>{@code
+		 * // See FabricPacket for creating the packet
+		 * ClientPlayNetworking.registerReceiver(OVERLAY_PACKET_TYPE, (player, packet, responseSender) -&rt; {
```
```suggestion
		 * ClientPlayNetworking.registerReceiver(OVERLAY_PACKET_TYPE, (player, packet, responseSender) -> {
```
1. I think it should be `&gt;` in the first place
2. `{@code}` escapes lt and gt anyway

Juuxel commented on fabric-networking-api-v1/src/client/java/net/fabricmc/fabric/api/client/networking/v1/ClientPlayNetworking.java:
```diff
@@ -255,4 +403,44 @@ public interface PlayChannelHandler {
+		 */
+		void receive(MinecraftClient client, ClientPlayerEntity player, T packet, PacketSender responseSender);
+
+		default boolean processOnNetworkThread(MinecraftClient client, ClientPlayerEntity player, T packet, PacketSender responseSender) {
```
Should have jd

Juuxel commented on fabric-networking-api-v1/src/main/java/net/fabricmc/fabric/api/networking/v1/ServerPlayNetworking.java:
```diff
@@ -197,6 +348,20 @@ public static boolean canSend(ServerPlayNetworkHandler handler, Identifier chann
+	 *
+	 * @param handler the network handler
+	 * @param type the packet type
+	 * @return True if the connected client has declared the ability to receive a specific type of packet
```
```suggestion
	 * @return {@code true} if the connected client has declared the ability to receive a specific type of packet
```

Juuxel commented on fabric-networking-api-v1/src/main/java/net/fabricmc/fabric/api/networking/v1/ServerPlayNetworking.java:
```diff
@@ -293,4 +475,56 @@ public interface PlayChannelHandler {
+		 * <p>An example usage of this is to create an explosion where the player is looking:
+		 * <pre>{@code
+		 * // See FabricPacket for creating the packet
+		 * ServerPlayNetworking.registerReceiver(BOOM_PACKET_TYPE, (player, packet, responseSender) -&rt; {
```
```suggestion
		 * ServerPlayNetworking.registerReceiver(BOOM_PACKET_TYPE, (player, packet, responseSender) -> {
```

Juuxel commented on fabric-networking-api-v1/src/client/java/net/fabricmc/fabric/api/client/networking/v1/ClientPlayNetworking.java:
```diff
@@ -41,6 +46,9 @@
  *
  * <p>This class should be only used on the physical client and for the logical client.
  *
+ * <p>See {@link ServerPlayNetworking} for information on how to use the new packet-object
+ * based API.
```
```suggestion
 * <p>See {@link ServerPlayNetworking} for information on how to use the packet object-based API.
```
It won't be new in the future, and I think the hyphen needs to be on "...-based"

apple502j committed:
Make PacketType#read public

apple502j committed:
Fix issues

liach commented on fabric-networking-api-v1/src/main/java/net/fabricmc/fabric/api/networking/v1/PacketType.java:
```diff
@@ -0,0 +1,75 @@
+ *
+ * @param <T> the type of the packet
+ */
+public final class PacketType<T extends FabricPacket> {
```
I want to move the packet write method to a function in the packet type as well, so we can use third-party java classes as packet types.

Technici4n replied:
That doesn't really make sense to me. And it's pretty easy to make a record to wrap a java object as a fabric packet.

apple502j replied:
Yeah, the vanilla-like system is more intuitive. (See also BlockEntity, Entity etc, where subclass does serialization)

Technici4n reviewed (commented):
I wonder if we shouldn't split this stuff to a separate class? 500 LOC for `ServerPlayNetworking` is a lot.

apple502j commented:
@Technici4n Not sure how to split those - the file is big, yes, but there isn't a clear way of splitting them.

If we're doing a v2, we can drop some methods that don't make sense in FAPI, but not right now.

apple502j commented:
Question: should there be Packet#getType? Could remove the packet type from send methods then.

Technici4n commented:
Sure, no split is probably fine.

> Question: should there be Packet#getType? Could remove the packet type from send methods then.

Not sure, either is probably ok.

Juuxel commented:
Having `getType` would match vanilla code where most things have that method, so it'd be a good change imo. I don't really see multiple types for one packet instance happening

apple502j committed:
Add FabricPacket#getType

modmuss50 commented on fabric-networking-api-v1/src/testmod/java/net/fabricmc/fabric/test/networking/play/NetworkingPlayPacketTest.java:
```diff
@@ -32,17 +32,16 @@
 	public static final Identifier TEST_CHANNEL = NetworkingTestmods.id("test_channel");
 
 	public static void sendToTestChannel(ServerPlayerEntity player, String stuff) {
-		PacketByteBuf buf = PacketByteBufs.create();
-		buf.writeText(Text.literal(stuff));
-		ServerPlayNetworking.send(player, TEST_CHANNEL, buf);
```
We should contiune to test the old API, its going no where.

apple502j replied:
Ok, reverting.

modmuss50 commented on fabric-networking-api-v1/src/client/java/net/fabricmc/fabric/api/client/networking/v1/ClientPlayNetworking.java:
```diff
@@ -255,4 +402,57 @@ public interface PlayChannelHandler {
+		 * @param responseSender the packet sender
+		 * @return whether to proceed to {@link #receive}
+		 */
+		default boolean processOnNetworkThread(MinecraftClient client, ClientPlayerEntity player, T packet, PacketSender responseSender) {
```
Maybe put the packet first?

apple502j replied:
The signature was from the classic API, which had the handler first. That said yeah, packet first makes more sense.

Technici4n replied:
Packet first makes more sense indeed.

modmuss50 commented on fabric-networking-api-v1/src/client/java/net/fabricmc/fabric/api/client/networking/v1/ClientPlayNetworking.java:
```diff
@@ -255,4 +402,57 @@ public interface PlayChannelHandler {
+		 * @param responseSender the packet sender
+		 * @return whether to proceed to {@link #receive}
+		 */
+		default boolean processOnNetworkThread(MinecraftClient client, ClientPlayerEntity player, T packet, PacketSender responseSender) {
```
Question: Do we want to pass some sort of context object instad of the client and player? Would make it easier to expand in the future without breaking?

Juuxel replied:
Also, the player is available from the client instance

apple502j replied:
I believe our convention is not to pass context object unless it makes more sense (see DynamicRegistrySetupCallback).

The reason this passes player is simple - it passed `player` only first, realized `player.client` is protected, so I passed client as well; `client.player` does exist, but it is marked `@Nullable` (when, here, it is never null). I believe the code here is fine.

Sollace replied:
Client is available from a static reference at `MinecraftClient.getInstance()`. I don't see any need for it to be passed at all.

apple502j replied:
Contributing guideline specifies, "Consider passing the MinecraftClient instance as a parameter if it makes sense."

Sollace replied:
Operative words being "if it makes sense". I explicitly ignore it in all of my networking code because it's just another client-exclusive class that I'd have to perform workarounds to reference in a way that won't break on the server.

Passing player makes sense, because I assume it's possible to receive packets for an instance of ClientOtherPlayerEntity as well (I hope), whereas passing client is never more than a convenience since it's always going to be accessible via the public non-nullable, static singleton reference (and is never not that).

modmuss50 commented on fabric-networking-api-v1/src/client/java/net/fabricmc/fabric/api/client/networking/v1/ClientPlayNetworking.java:
```diff
@@ -255,4 +402,57 @@ public interface PlayChannelHandler {
+		void receive(MinecraftClient client, ClientPlayerEntity player, T packet, PacketSender responseSender);
+
+		/**
+		 * Processes an incoming packet. This is called on the network thread.
+		 * If this method returns {@code false}, {@link #receive} is not called.
+		 * This is for advanced users only; for normal uses, {@link #receive} should be
+		 * implemented instead.
```
Add a test for this behaviour please.

modmuss50 commented on fabric-networking-api-v1/src/client/java/net/fabricmc/fabric/api/client/networking/v1/ClientPlayNetworking.java:
```diff
@@ -255,4 +402,57 @@ public interface PlayChannelHandler {
+		 *
+		 * <p>An example usage of this is to display an overlay message:
+		 * <pre>{@code
+		 * // See FabricPacket for creating the packet
```
Nit: is this code comment really needed?

apple502j replied:
Well, it's never a bad idea to do over-documenting 😛

modmuss50 commented on fabric-networking-api-v1/src/client/java/net/fabricmc/fabric/api/client/networking/v1/ClientPlayNetworking.java:
```diff
@@ -62,6 +76,39 @@ public static boolean registerGlobalReceiver(Identifier channelName, PlayChannel
+	 * @see ClientPlayNetworking#registerReceiver(PacketType, PlayPacketHandler)
+	 */
+	public static <T extends FabricPacket> boolean registerGlobalReceiver(PacketType<T> type, PlayPacketHandler<T> handler) {
+		return registerGlobalReceiver(type.getId(), new PlayChannelHandlerProxy<T>() {
+			@Override
+			public PlayPacketHandler<T> getOriginalHandler() {
+				return handler;
+			}
+
+			@Override
+			public void receive(MinecraftClient client, ClientPlayNetworkHandler networkHandler, PacketByteBuf buf, PacketSender sender) {
+				T packet = type.read(buf);
+
+				if (handler.processOnNetworkThread(client, client.player, packet, sender)) {
+					client.execute(() -> {
+						if (networkHandler.getConnection().isOpen()) handler.receive(client, client.player, packet, sender);
+					});
+				}
+			}
+		});
```
Nit: maybe move this to an impl class, and possibly remove duplicatated logic with registerReceiver?

apple502j replied:
Maybe, want more opinions on this

Sollace replied:
+1 on removing duplicated logic

modmuss50 reviewed (changes requested):
This is looking great, just a few small things 👍

apple502j committed:
Merge branch '1.19.3' of https://github.com/FabricMC/fabric into networking-api/new-packet-handler

Technici4n commented on fabric-networking-api-v1/src/client/java/net/fabricmc/fabric/api/client/networking/v1/ClientPlayNetworking.java:
```diff
@@ -49,9 +57,15 @@ public final class ClientPlayNetworking {
 	 * Use {@link #unregisterGlobalReceiver(Identifier)} to unregister the existing handler.
 	 *
+	 * <p>For new code, {@link #registerGlobalReceiver(PacketType, PlayPacketHandler)}
+	 * should be used instead, as it is designed in a way that prevents thread safety issues.
```
```suggestion
	 * should be preferred, as it is designed in a way that prevents thread safety issues.
```

Technici4n replied:
Unless the network thread processing can do everything that the previous API can do? (In which case it's fine to deprecate the old API).

Technici4n commented on fabric-networking-api-v1/src/client/java/net/fabricmc/fabric/api/client/networking/v1/ClientPlayNetworking.java:
```diff
@@ -49,9 +57,15 @@ public final class ClientPlayNetworking {
 	 * Registers a handler to a channel.
 	 * A global receiver is registered to all connections, in the present and future.
 	 *
+	 * <p>The handler runs on the network thread. After reading the buffer there, rendering
+	 * must be performed in the render thread by calling {@link ThreadExecutor#execute(Runnable)}.
```
```suggestion
	 * <p>The handler runs on the network thread. After reading the buffer there, access to
	 * game state must be performed in the render thread by calling {@link ThreadExecutor#execute(Runnable)}.
```

Technically it's not just rendering.

Technici4n commented on fabric-networking-api-v1/src/client/java/net/fabricmc/fabric/api/client/networking/v1/ClientPlayNetworking.java:
```diff
@@ -97,6 +163,9 @@ public static Set<Identifier> getGlobalReceivers() {
 	 * login query has been received, you should use {@link ClientPlayConnectionEvents#INIT} to register the channel handler.
 	 *
+	 * <p>For new code, {@link #registerReceiver(PacketType, PlayPacketHandler)}
+	 * should be used instead, as it is designed in a way that prevents thread safety issues.
```
```suggestion
	 * should be preferred, as it is designed in a way that prevents thread safety issues.
```

apple502j replied:
We prefer, not the readers; changed to "is preferred"

Technici4n commented on fabric-networking-api-v1/src/client/java/net/fabricmc/fabric/api/client/networking/v1/ClientPlayNetworking.java:
```diff
@@ -255,4 +402,57 @@ public interface PlayChannelHandler {
+		 * @param player the player that received the packet
+		 * @param packet the packet
+		 * @param responseSender the packet sender
+		 * @return whether to proceed to {@link #receive}
```
This is unclear and needs to specify explicitly what happens with `{@code true}` / `{@code false}`. Even then, the name could be misleading (one could understand it as "should processing happen on network thread"). A less ambiguous name would be nice.

apple502j committed:
Restore tests for traditional packet handling

apple502j committed:
Change method signatures

apple502j committed:
Javadoc nitpick

apple502j committed:
Improve processOnNetworkThread test

Juuxel commented on fabric-networking-api-v1/src/testmod/java/net/fabricmc/fabric/test/networking/channeltest/NetworkingChannelTest.java:
```diff
@@ -94,6 +94,7 @@ private static CompletableFuture<Suggestions> suggestReceivableChannels(CommandC
 		return CommandSource.suggestIdentifiers(ServerPlayNetworking.getReceived(player), builder);
 	}
 
+	@SuppressWarnings("deprecation")
```
Was anything here actually deprecated?

apple502j replied:
Nope, removing

Juuxel commented on fabric-networking-api-v1/src/testmod/java/net/fabricmc/fabric/test/networking/channeltest/NetworkingChannelTest.java:
```diff
@@ -110,6 +111,7 @@ private static int registerChannel(CommandContext<ServerCommandSource> context,
 		return 1;
 	}
 
+	@SuppressWarnings("deprecation")
```
same here

apple502j committed:
Remove unused annotation

modmuss50 commented on fabric-networking-api-v1/src/client/java/net/fabricmc/fabric/api/client/networking/v1/ClientPlayNetworking.java:
```diff
@@ -255,4 +402,56 @@ public interface PlayChannelHandler {
+		 * @param responseSender the packet sender
+		 * @return {@code true} to proceed to {@link #receive}
+		 */
+		default boolean processOnNetworkThread(T packet, ClientPlayerEntity player, PacketSender responseSender) {
```
Im a little unsure about the naming for this? Its fine unless someone can come up with something better.

Technici4n replied:
I agree it should be changed, the name is the only blocker for me.

apple502j replied:
Hmm, yeah, we need a better name but I can't really come up with one

apple502j committed:
Merge branch '1.19.3' of https://github.com/FabricMC/fabric into networking-api/new-packet-handler

apple502j committed:
Rename processOnNetworkThread

apple502j commented:
Removing from last call now, we might want to remove network thread thing after all

apple502j committed:
Remove network-thread handling

apple502j committed:
Do not submit when processing on main thread

In 1.19.4, S2C packets can be processed on the main thread through packet bundling.
While 1.19.3 does not have this feature, and C2S packets cannot be bundled, there is
no harm doing it right now.

apple502j committed:
Small change

apple502j commented:
This is last-call ready.

modmuss50 commented:
Manually merged into 1.19.4, Many Thanks 👍
